message_list_tooltips: Convert module to TypeScript.

LaPulgaaa opened the pull request:
<!-- Describe your pull request here.-->

LaPulgaaa commented on web/src/message_list_tooltips.ts:
```diff
@@ -82,7 +82,9 @@ function hide_tooltip_if_reference_removed(
         // We have to be smart about hiding the instance, so we hide it as soon
         // as it is displayed.
-        setTimeout(instance.hide, 0);
+        setTimeout(() => {
```
[eslint-rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/unbound-method.mdx)

LaPulgaaa commented on web/src/message_list_tooltips.ts:
```diff
@@ -132,16 +139,19 @@ export function initialize() {
-            const message_id = rows.get_message_id(instance.reference);
+            assert(local_id !== undefined);
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            const message_id = rows.get_message_id(instance.reference as HTMLElement);
```
The function parameter  type is conflicting with `tippyjs` type upstream. Can we do something else here ?

timabbott replied:
It's better to just do an assert that `instance.reference` is an HTMLElement than do an untyped cast; that'll avoid the need for an `eslint` disable here.

LaPulgaaa commented on web/src/message_list_tooltips.ts:
```diff
@@ -78,7 +85,7 @@ function hide_tooltip_if_reference_removed(
         return;
     }
-    const callback = function (mutationsList) {
+    const callback = function (mutationsList: MutationRecord[]): void {
```
Reference: https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.mutationrecord.html

LaPulgaaa commented:
There are 3 instances where the upstream `tippyjs` type has conflicted with local function parameter type. I have `type-assert` ed those `variables/parameters` but am looking for some advice if we can handle it differently.

timabbott commented on web/src/message_list_tooltips.js:
```diff
@@ -266,19 +266,22 @@ export function initialize() {
-        onHidden(instance) {
-            instance.destroy();
+    message_list_tooltip(
+        [".rendered_markdown time", ".rendered_markdown .copy_codeblock"].join(","),
```
Why is this required? I wonder if we just need to declare the type on `message_list_tooltip` better to pass-through the actual upstream type that supports string or array.

LaPulgaaa replied:
This acts as `target` argument when calling  `tippyjs` `delegate` function. The `delegate` function accepts argument of type `tippy.Props` which has the `target` field of type `string`. Reference:https://github.com/atomiks/tippyjs/blob/ad85f6feb79cf6c5853c43bf1b2a50c4fa98e7a1/src/types.ts#L171

timabbott replied:
OK, but this seems like a better solution, applied:

``` diff
diff --git a/web/src/message_list_tooltips.js b/web/src/message_list_tooltips.js
index 464fee44a2..ce2c05fc1c 100644
--- a/web/src/message_list_tooltips.js
+++ b/web/src/message_list_tooltips.js
@@ -273,8 +273,7 @@ export function initialize() {
         },
     });
 
-    message_list_tooltip(
-        [".rendered_markdown time", ".rendered_markdown .copy_codeblock"].join(","),
+    message_list_tooltip(".rendered_markdown time, .rendered_markdown .copy_codeblock",
         {
             content: timerender.get_markdown_time_tooltip,
             onHidden(instance) {
```

LaPulgaaa replied:
Noted. Thanks !.

timabbott commented on web/src/message_list_tooltips.ts:
```diff
@@ -193,7 +204,7 @@ export function initialize() {
             const tippy_content = $elem.attr("data-tippy-content");
-            const $template = $(`#${CSS.escape($elem.attr("data-tooltip-template-id"))}`);
+            const $template = $(`#${CSS.escape($elem.attr("data-tooltip-template-id")!)}`);
             instance.setContent(tippy_content ?? parse_html($template.html()));
```
I think the `!` might not be right here; I think we only expect `data-tooltip-template-id` to be present if `tippy_content` is not valid. Might need a prep commit to rewrite this logic to not generate an invalid `$template` in the event that `data-tippy-content` was present, using if/else clauses or something.

timabbott commented on web/src/message_list_tooltips.ts:
```diff
@@ -276,7 +289,8 @@ export function initialize() {
         {
-            content: timerender.get_markdown_time_tooltip,
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            content: timerender.get_markdown_time_tooltip as tippy.Content,
```
This may be annoying to avoid the cast; it's expecting a function with type `((ref: Element) => string | Element | DocumentFragment);`, and our function has type `HTMLElement => string | DocumentFragment`.

LaPulgaaa replied:
As you stated `timerender.get_markdown_time_tooltip` is of type `HTMLElement => string | DocumentFragment` but it is expecting `content` parameter of type `tippy.Content` (an optional field of upstream `tippy.Props` ).

timabbott replied:
Yeah; that Tippy type is a union, so it could work without the cast if not for the `Element`/`HTMLElement` distinction, I think.

LaPulgaaa replied:
Yes, I agree!

LaPulgaaa commented on web/src/message_list_tooltips.ts:
```diff
@@ -226,9 +238,10 @@ export function initialize() {
             const $time_elem = $(instance.reference);
-            const $row = $time_elem.closest(".message_row");
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            const $row = $time_elem.closest(".message_row") as JQuery;
```
To be precise `$time_elem` is of type `JQuery<tippy.ReferenceElement<tippy.Props>>` as instance is of type `tippy.Instance`. Hence the code block 
```
            const $row = $time_elem.closest(".message_row") as JQuery;
```
returns `$row` as type `JQuery<tippy.ReferenceElement<tippy.Props>>` and `rows.id($row)` expects `$row` to be of type `JQuery<HTMLElement>`

timabbott replied:
I wonder whether doing `div.message_row` as the selector would help? But probably not worth spending a ton of time on it.

timabbott commented:
@LaPulgaaa can you post an update on this? I think some of the comments above may not require action, but at least
https://github.com/zulip/zulip/pull/29335#discussion_r1529522986 seems to suggest you intend to make a change. Would be nice to integrate this.

LaPulgaaa commented:
Apologies for any delay. I have added a prep commit and dropped a comment addressing [#29335(comment)](https://github.com/zulip/zulip/pull/29335#discussion_r1529522986)

timabbott commented:
The test failure is an operational issue: https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/debian.2012.20gpg.20failures/near/1764206

Merged, after squashing the linter-fixup commit and the small tweak noted above, thanks @LaPulgaaa!
